File: android/app/src/main/kotlin/com/poracode/app/chat/TerminalCursorReconciler.kt
package com.poracode.app.chat

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

enum class TerminalCursorFrameKind {
    BASELINE,
    OUTPUT,
}

data class TerminalCursorFrame(
    val kind: TerminalCursorFrameKind,
    val terminalId: String,
    val watchId: String,
    val generation: String?,
    val fromCursor: Long,
    val toCursor: Long,
    val data: String,
)

data class TerminalCursorState(
    val watchId: String,
    val baselineReceived: Boolean = false,
    val generation: String? = null,
    val toCursor: Long = 0L,
    /** Bounded tail. Cursor coordinates remain absolute even after trimming. */
    val transcript: String = "",
    val bufferedOutput: List<TerminalCursorFrame> = emptyList(),
    val bufferedUtf16Units: Int = 0,
    val needsResync: Boolean = false,
) {
    init {
        require(watchId.isNotEmpty()) { "watchId must not be empty" }
        require(toCursor >= 0L) { "toCursor must not be negative" }
        require(transcript.length <= TerminalCursorReconciler.MAX_TRANSCRIPT_UTF16_UNITS)
    }

    companion object {
        fun watching(watchId: String): TerminalCursorState = TerminalCursorState(watchId)

        fun established(
            watchId: String,
            generation: String?,
            toCursor: Long,
            transcript: String = "",
        ): TerminalCursorState = TerminalCursorState(
            watchId = watchId,
            baselineReceived = true,
            generation = generation,
            toCursor = toCursor,
            transcript = transcript.takeLast(TerminalCursorReconciler.MAX_TRANSCRIPT_UTF16_UNITS),
        )
    }
}

enum class TerminalCursorAction {
    BUFFER,
    REPLACE,
    IGNORE,
    APPEND,
    APPEND_UNSEEN_SUFFIX,
    RESYNC,
}

data class TerminalCursorResult(
    val state: TerminalCursorState,
    val action: TerminalCursorAction,
    val appendedText: String = "",
)

object TerminalCursorFrameDecoder {
    fun decode(value: JsonElement): TerminalCursorFrame? {
        val objectValue = value.objectOrNull() ?: return null
        val terminalId = objectValue.requiredString("id", allowEmpty = false) ?: return null
        return when (objectValue.requiredString("type")) {
            "terminal-output" -> {
                val data = objectValue.requiredString("data") ?: return null
                val sync = objectValue["cursorSync"]?.objectOrNull() ?: return null
                val generation = sync.requiredString("generation", allowEmpty = false) ?: return null
                decodeRange(TerminalCursorFrameKind.OUTPUT, terminalId, generation, data, sync)
            }
            "terminal-watch-result" -> {
                val sync = objectValue["cursorSync"]?.objectOrNull() ?: return null
                if (sync["version"]?.longOrStrictNull() != 1L) return null
                val watchId = sync.requiredString("watchId", allowEmpty = false) ?: return null
                val result = sync["result"]?.objectOrNull() ?: return null
                if (result.requiredString("status") != "ready") return null
                val generation = when (val raw = result["generation"]) {
                    JsonNull -> null
                    else -> raw?.stringOrNull() ?: return null
                }
                val data = result.requiredString("data") ?: return null
                val from = result["fromCursor"]?.longOrStrictNull() ?: return null
                val to = result["toCursor"]?.longOrStrictNull() ?: return null
                TerminalCursorFrame(
                    TerminalCursorFrameKind.BASELINE,
                    terminalId,
                    watchId,
                    generation,
                    from,
                    to,
                    data,
                ).takeIf(::validRange)
            }
            else -> null
        }
    }

    private fun decodeRange(
        kind: TerminalCursorFrameKind,
        terminalId: String,
        generation: String?,
        data: String,
        sync: kotlinx.serialization.json.JsonObject,
    ): TerminalCursorFrame? {
        if (sync["version"]?.longOrStrictNull() != 1L) return null
        val watchId = sync.requiredString("watchId", allowEmpty = false) ?: return null
        val from = sync["fromCursor"]?.longOrStrictNull() ?: return null
        val to = sync["toCursor"]?.longOrStrictNull() ?: return null
        return TerminalCursorFrame(kind, terminalId, watchId, generation, from, to, data)
            .takeIf(::validRange)
    }

    private fun validRange(frame: TerminalCursorFrame): Boolean =
        frame.fromCursor >= 0L &&
            frame.toCursor >= frame.fromCursor &&
            frame.toCursor - frame.fromCursor == frame.data.length.toLong()
}

object TerminalCursorReconciler {
    const val MAX_TRANSCRIPT_UTF16_UNITS: Int = 200_000
    private const val MAX_BUFFERED_UTF16_UNITS: Int = 200_000
    private const val MAX_BUFFERED_FRAMES: Int = 1_024

    fun reconcile(
        state: TerminalCursorState,
        frame: TerminalCursorFrame,
    ): TerminalCursorResult {
        if (frame.watchId != state.watchId) {
            return TerminalCursorResult(state, TerminalCursorAction.IGNORE)
        }
        if (!validRange(frame)) return resync(state)
        if (frame.kind == TerminalCursorFrameKind.BASELINE) return replaceBaseline(state, frame)
        if (!state.baselineReceived) return bufferBeforeBaseline(state, frame)
        if (state.needsResync) return resync(state)
        return appendOutput(state, frame)
    }

    private fun replaceBaseline(
        state: TerminalCursorState,
        frame: TerminalCursorFrame,
    ): TerminalCursorResult {
        var next = state.copy(
            baselineReceived = true,
            generation = frame.generation,
            toCursor = frame.toCursor,
            transcript = boundedTail(frame.data),
            bufferedOutput = emptyList(),
            bufferedUtf16Units = 0,
            needsResync = false,
        )
        if (frame.generation != null) {
            for (buffered in state.bufferedOutput) {
                val result = appendOutput(next, buffered)
                next = result.state
                if (result.action == TerminalCursorAction.RESYNC) break
            }
        }
        return TerminalCursorResult(next, TerminalCursorAction.REPLACE)
    }

    private fun bufferBeforeBaseline(
        state: TerminalCursorState,
        frame: TerminalCursorFrame,
    ): TerminalCursorResult {
        if (frame.generation == null) return resync(state)
        val units = state.bufferedUtf16Units.toLong() + frame.data.length.toLong()
        if (units > MAX_BUFFERED_UTF16_UNITS ||
            state.bufferedOutput.size >= MAX_BUFFERED_FRAMES
        ) {
            return resync(
                state.copy(bufferedOutput = emptyList(), bufferedUtf16Units = 0),
            )
        }
        return TerminalCursorResult(
            state.copy(
                bufferedOutput = state.bufferedOutput + frame,
                bufferedUtf16Units = units.toInt(),
            ),
            TerminalCursorAction.BUFFER,
        )
    }

    private fun appendOutput(
        state: TerminalCursorState,
        frame: TerminalCursorFrame,
    ): TerminalCursorResult {
        if (state.generation == null || frame.generation == null) return resync(state)
        if (state.generation != frame.generation) return resync(state)
        if (frame.toCursor <= state.toCursor) {
            return TerminalCursorResult(state, TerminalCursorAction.IGNORE)
        }
        if (frame.fromCursor > state.toCursor) return resync(state)
        val overlap = (state.toCursor - frame.fromCursor).coerceAtLeast(0L)
        if (overlap > frame.data.length.toLong()) return resync(state)
        val suffix = frame.data.substring(overlap.toInt())
        val action = if (overlap == 0L) {
            TerminalCursorAction.APPEND
        } else {
            TerminalCursorAction.APPEND_UNSEEN_SUFFIX
        }
        val next = state.copy(
            toCursor = frame.toCursor,
            transcript = appendBounded(state.transcript, suffix),
        )
        return TerminalCursorResult(next, action, suffix)
    }

    private fun resync(state: TerminalCursorState): TerminalCursorResult =
        TerminalCursorResult(state.copy(needsResync = true), TerminalCursorAction.RESYNC)

    private fun validRange(frame: TerminalCursorFrame): Boolean =
        frame.fromCursor >= 0L &&
            frame.toCursor >= frame.fromCursor &&
            frame.toCursor - frame.fromCursor == frame.data.length.toLong()

    private fun boundedTail(value: String): String =
        if (value.length <= MAX_TRANSCRIPT_UTF16_UNITS) value
        else value.substring(value.length - MAX_TRANSCRIPT_UTF16_UNITS)

    private fun appendBounded(previous: String, suffix: String): String {
        if (suffix.length >= MAX_TRANSCRIPT_UTF16_UNITS) return boundedTail(suffix)
        val keepPrevious = MAX_TRANSCRIPT_UTF16_UNITS - suffix.length
        return boundedTail(previous).takeLast(keepPrevious) + suffix
    }
}

File: ios/App/App/Features/RichChat/TerminalCursorReconciler.swift
import Foundation

enum TerminalCursorFrameKind: Sendable, Equatable {
  case baseline
  case output
}

struct TerminalCursorFrame: Sendable, Equatable {
  let kind: TerminalCursorFrameKind
  let terminalID: String
  let watchID: String
  let generation: String?
  let fromCursor: Int64
  let toCursor: Int64
  let data: String
}

struct TerminalCursorPosition: Sendable, Equatable {
  let generation: String?
  let toCursor: Int64
}

struct TerminalCursorState: Sendable, Equatable {
  let watchID: String
  var baselineReceived = false
  var generation: String?
  var toCursor: Int64 = 0
  /// Bounded display tail; cursor positions remain absolute after trimming.
  var transcript = ""
  var bufferedOutput: [TerminalCursorFrame] = []
  var bufferedUTF16Units = 0
  var needsResync = false

  static func watching(_ watchID: String) -> Self { Self(watchID: watchID) }

  static func established(
    watchID: String,
    generation: String?,
    toCursor: Int64,
    transcript: String = ""
  ) -> Self {
    Self(
      watchID: watchID,
      baselineReceived: true,
      generation: generation,
      toCursor: toCursor,
      transcript: TerminalCursorReconciler.boundedTail(transcript),
      bufferedOutput: [],
      bufferedUTF16Units: 0,
      needsResync: false
    )
  }
}

enum TerminalCursorAction: String, Sendable, Equatable {
  case buffer
  case replace
  case ignore
  case append
  case appendUnseenSuffix = "append-unseen-suffix"
  case resync
}

enum TerminalCursorReconciliationReason: Sendable, Equatable {
  case staleWatch
  case invalidRange
  case missingBaseline
  case generationChanged
  case cursorGap
  case invalidUTF16Boundary
}

struct TerminalCursorResult: Sendable, Equatable {
  let state: TerminalCursorState
  let action: TerminalCursorAction
  let appendedText: String
  let reason: TerminalCursorReconciliationReason?

  var isStaleWatch: Bool { reason == .staleWatch }
}

enum TerminalCursorFrameDecoder {
  static func decode(_ value: RichJSON) throws -> TerminalCursorFrame {
    guard let object = value.objectValue,
      let type = RichDecoding.requiredString(object, "type"),
      let terminalID = RichDecoding.requiredString(object, "id", allowEmpty: false)
    else { throw RichDomainDecodeError.invalidTerminalFrame }
    if type == "terminal-output" {
      guard let data = RichDecoding.requiredString(object, "data"),
        let sync = object["cursorSync"]?.objectValue,
        let generation = RichDecoding.requiredString(sync, "generation", allowEmpty: false)
      else { throw RichDomainDecodeError.invalidTerminalFrame }
      return try decodeRange(
        kind: .output,
        terminalID: terminalID,
        generation: generation,
        data: data,
        sync: sync
      )
    }
    if type == "terminal-watch-result" {
      guard let sync = object["cursorSync"]?.objectValue,
        sync["version"]?.exactInt64Value == 1,
        let watchID = RichDecoding.requiredString(sync, "watchId", allowEmpty: false),
        let result = sync["result"]?.objectValue,
        RichDecoding.requiredString(result, "status") == "ready",
        let data = RichDecoding.requiredString(result, "data"),
        let from = result["fromCursor"]?.exactInt64Value,
        let to = result["toCursor"]?.exactInt64Value,
        result.keys.contains("generation")
      else { throw RichDomainDecodeError.invalidTerminalFrame }
      let generation: String?
      if result["generation"] == .null {
        generation = nil
      } else if let value = result["generation"]?.stringValue, !value.isEmpty {
        generation = value
      } else {
        throw RichDomainDecodeError.invalidTerminalFrame
      }
      let frame = TerminalCursorFrame(
        kind: .baseline,
        terminalID: terminalID,
        watchID: watchID,
        generation: generation,
        fromCursor: from,
        toCursor: to,
        data: data
      )
      guard TerminalCursorReconciler.isValid(frame) else {
        throw RichDomainDecodeError.invalidTerminalFrame
      }
      return frame
    }
    throw RichDomainDecodeError.invalidTerminalFrame
  }

  private static func decodeRange(
    kind: TerminalCursorFrameKind,
    terminalID: String,
    generation: String?,
    data: String,
    sync: [String: RichJSON]
  ) throws -> TerminalCursorFrame {
    guard sync["version"]?.exactInt64Value == 1,
      let watchID = RichDecoding.requiredString(sync, "watchId", allowEmpty: false),
      let from = sync["fromCursor"]?.exactInt64Value,
      let to = sync["toCursor"]?.exactInt64Value
    else { throw RichDomainDecodeError.invalidTerminalFrame }
    let frame = TerminalCursorFrame(
      kind: kind,
      terminalID: terminalID,
      watchID: watchID,
      generation: generation,
      fromCursor: from,
      toCursor: to,
      data: data
    )
    guard TerminalCursorReconciler.isValid(frame) else {
      throw RichDomainDecodeError.invalidTerminalFrame
    }
    return frame
  }
}

enum TerminalCursorReconciler {
  static let maximumTranscriptUTF16Units = 200_000
  private static let maximumBufferedUTF16Units = 200_000

  static func isStale(frame: TerminalCursorFrame, currentWatchID: String) -> Bool {
    frame.watchID != currentWatchID
  }

  static func isAppendCompatible(
    previous: TerminalCursorPosition?, frame: TerminalCursorFrame
  ) -> Bool {
    guard let previous else { return true }
    return previous.generation == frame.generation && previous.toCursor == frame.fromCursor
  }

  static func reconcile(
    state: TerminalCursorState,
    frame: TerminalCursorFrame
  ) -> TerminalCursorResult {
    guard frame.watchID == state.watchID else {
      return result(state, .ignore, reason: .staleWatch)
    }
    guard isValid(frame) else { return resync(state, reason: .invalidRange) }
    if frame.kind == .baseline { return replaceBaseline(state, frame) }
    if !state.baselineReceived { return bufferBeforeBaseline(state, frame) }
    if state.needsResync { return resync(state, reason: .missingBaseline) }
    return appendOutput(state, frame)
  }

  static func isValid(_ frame: TerminalCursorFrame) -> Bool {
    frame.fromCursor >= 0 && frame.toCursor >= frame.fromCursor
      && frame.toCursor - frame.fromCursor == Int64(frame.data.utf16.count)
  }

  static func boundedTail(_ value: String) -> String {
    let count = value.utf16.count
    guard count > maximumTranscriptUTF16Units else { return value }
    var dropped = count - maximumTranscriptUTF16Units
    while dropped <= count {
      if let suffix = suffix(value, droppingUTF16Units: dropped) { return suffix }
      dropped += 1
    }
    return ""
  }

  private static func replaceBaseline(
    _ state: TerminalCursorState, _ frame: TerminalCursorFrame
  ) -> TerminalCursorResult {
    var next = TerminalCursorState.established(
      watchID: state.watchID,
      generation: frame.generation,
      toCursor: frame.toCursor,
      transcript: frame.data
    )
    if frame.generation != nil {
      for buffered in state.bufferedOutput {
        let replay = appendOutput(next, buffered)
        next = replay.state
        if replay.action == .resync { break }
      }
    }
    return result(next, .replace)
  }

  private static func bufferBeforeBaseline(
    _ state: TerminalCursorState, _ frame: TerminalCursorFrame
  ) -> TerminalCursorResult {
    guard frame.generation != nil else { return resync(state, reason: .missingBaseline) }
    let added = frame.data.utf16.count
    guard state.bufferedUTF16Units <= maximumBufferedUTF16Units - added else {
      var cleared = state
      cleared.bufferedOutput = []
      cleared.bufferedUTF16Units = 0
      return resync(cleared, reason: .missingBaseline)
    }
    var next = state
    next.bufferedOutput.append(frame)
    next.bufferedUTF16Units += added
    return result(next, .buffer)
  }

  private static func appendOutput(
    _ state: TerminalCursorState, _ frame: TerminalCursorFrame
  ) -> TerminalCursorResult {
    guard let generation = state.generation, frame.generation == generation else {
      return resync(state, reason: .generationChanged)
    }
    if frame.toCursor <= state.toCursor { return result(state, .ignore) }
    guard frame.fromCursor <= state.toCursor else { return resync(state, reason: .cursorGap) }
    let overlap = state.toCursor - frame.fromCursor
    guard overlap <= Int64(frame.data.utf16.count),
      let suffix = suffix(frame.data, droppingUTF16Units: Int(overlap))
    else { return resync(state, reason: .invalidUTF16Boundary) }
    var next = state
    next.toCursor = frame.toCursor
    next.transcript = boundedTail(state.transcript + suffix)
    return result(next, overlap == 0 ? .append : .appendUnseenSuffix, appended: suffix)
  }

  private static func suffix(_ value: String, droppingUTF16Units count: Int) -> String? {
    guard count >= 0, count <= value.utf16.count else { return nil }
    let utf16Index = value.utf16.index(value.utf16.startIndex, offsetBy: count)
    guard let index = String.Index(utf16Index, within: value) else { return nil }
    return String(value[index...])
  }

  private static func resync(
    _ state: TerminalCursorState, reason: TerminalCursorReconciliationReason
  ) -> TerminalCursorResult {
    var next = state
    next.needsResync = true
    return result(next, .resync, reason: reason)
  }

  private static func result(
    _ state: TerminalCursorState,
    _ action: TerminalCursorAction,
    appended: String = "",
    reason: TerminalCursorReconciliationReason? = nil
  ) -> TerminalCursorResult {
    TerminalCursorResult(state: state, action: action, appendedText: appended, reason: reason)
  }
}
